base/array.jl

```diff
@@ -1532,7 +1532,7 @@ julia> findnext(isodd, A, 2)
 function findnext(testf::Function, A, start::Integer)
     l = endof(A)
     i = start
-    while i <= l
+    @inbounds while i <= l
```

Comment:
This will apply `@inbounds` to the user-provided `testf` as well.
There are also a few cases like this below.
For the Base it's better to use `@inbounds` really "atomically".

Reply:
>This will apply @inbounds to the user-provided testf as well.
Is this bad?
>For the Base it's better to use @inbounds really "atomically".
Ok, I was not aware of this coding standard (I thought the code looks nicer if we pull out the inbounds as far as possible, instead of replicating it inside).

I'll change to comply. For cases `a[i] = b[j]`, is there a preferred way of annotating inbounds separately for the getindex and setindex! ? The obvious way is to write `bj=b[j];a[i]=bj` and then separately annotate it; but this looks somewhat ugly.

Reply:
> Is this bad?

It's quite bad, because you are applying `@inbounds` to an unknown code. E.g. if it's fresh untested code, it'll crash Julia session.

> I thought the code looks nicer if we pull out the inbounds as far as possible

Then it's easy to overlook. If the code within the loop gets modified, the `@inbounds` might no longer apply safely.
"Fine-grained" `@inbounds` might look pedantic in the user code, but Base has higher standards.

> For cases a[i] = b[j], is there a preferred way of annotating inbounds separately for the getindex and setindex! ?

If from the context it follows that both `a[i]` and `b[j]` accesses are safe, you can just do `@inbounds a[i] = b[j]`, of course.

Reply:
See discussion at https://github.com/JuliaLang/julia/pull/24335.